cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -165,68 +172,99 @@ func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
 // Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
 // TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
 // matching prefix, but instance doesn't belong to it.
-func (gc *GceCache) GetMigForInstance(instance *GceRef) (Mig, error) {
+func (gc *GceCache) GetMigForInstance(instanceRef GceRef) (Mig, error) {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
 
-	if mig, found := gc.instancesCache[*instance]; found {
-		return mig, nil
+	if migRef, found := gc.instanceRefToMigRef[instanceRef]; found {
+		mig, found := gc.getMig(migRef)
+		if !found {
+			// It is fine that mig is not found in gc.migs map even if we have entry in gc.instanceRefToMigRef
+			// The mig could have been removed from gc.migs if we (for example) disabled autoscaling for it or the node pool
+			// was just deleted. Yet we are not removing entries from gc.instanceRefToMigRef in such case due to performance reasons.
```

Comment:
What performance reasons are those? Feels like you can easily clean up the map in UnregisterMig. Sure, you'd need to iterate over the whole map, but that should still be negligible.